### PR TITLE
test(scanner): Check scan and visit methods in CtInheritanceScanner.

### LINKED ISSUE #276: add implementation of CtInheritanceScanner contract
follow up of #269 

>  you SHOULD have one visitXXX method CtVisitor if the element could be instanced, OR a scanXXX method in CtInheritanceScanner if it's an abstract concept.

### PR DESCRIPTION
**Pull request by GerardPaligot**
Closes #276

### CODE REVIEW

**Comment by GerardPaligot** [comment]
@monperrus PR ok.

**Comment by monperrus** [comment]
very interesting PR

now we need a clear contract for CtVisitor, because it's in the hierarchy.
Should it be
declares write* methods for and only for factory instantiable metamodel classes?
declares write* methods for and only for factory non abstract metamodel classes?
(should those two contracts be equivalent?)

we also need a clear contract for CtScanner and AbstractVisitor, but this is independent of this issue and PR.
...
unless, CtInheritanceScanner should extend AbstractVisitor and not directly implement CtVisitor. This may be the case since they both declare scan(CtElement) and the like.

what do you think?

**Comment by GerardPaligot** [comment]
> now we need a clear contract for CtVisitor, because it's in the hierarchy.
> Should it be
> declares write* methods for and only for factory instantiable metamodel classes?
> declares write* methods for and only for factory non abstract metamodel classes?
> (should those two contracts be equivalent?)

I don't understand. What is `write*` methods? Do you mean `visit*`? Why do you talk about factory?

> unless, CtInheritanceScanner should extend AbstractVisitor and not directly implement CtVisitor. This may be the case since they both declare scan(CtElement) and the like.

Why not.

**Comment by monperrus** [comment]
> What is |write*| methods? Do you mean |visit*|? 
yes

Do you agree with the contract of CtVisitor:
'declares visit* methods for and only for non abstract metamodel classes'

--Martin

**Comment by GerardPaligot** [comment]
> Do you agree with the contract of CtVisitor: 'declares visit* methods for and only for non abstract metamodel classes'

Yes but this test is already done with `visit*` methods on `CtInheritanceScanner` class.

**Comment by monperrus** [comment]
it seems that this test is done at the level of CtInheritanceScanner not
at the level of CtVisitor.

If it is done at the CtVisitor level, we can remove the check from
CtInheritanceScanner since the compiler enforces it.

**Comment by GerardPaligot** [comment]
To resume, I remove all checks about `visit*` methods on `CtInheritanceScanner` and I create a new test case for `CtVisitor` with these checks. We are okay?

**Comment by monperrus** [comment]
> To resume, I remove all checks about |visit*| methods on
> |CtInheritanceScanner| and I create a new test case for |CtVisitor|
> with these checks. We are okay?
>
remove all checks about about the presence of
but keep the checks about the call of those methods.

--Martin

**Comment by GerardPaligot** [comment]
Do you talk about my test case or the test case of Nicolas with `CtInheritanceScannerTest`? Because, in my test case, I don't have checks about the call of `visit*` methods.

**Comment by monperrus** [comment]
That's right, it's another test. So only I guess it's only splitting your test in two, one for CtVisitor and one for  CtInheritanceScanner

By the way, could you add the contract in CtVisitor? and a couple of method Javadoc is missing there.

**Comment by GerardPaligot** [comment]
> and a couple of method Javadoc is missing there.

Where?

**Comment by monperrus** [comment]
just open the file :-)

    <T> void visitCtAnnotationFieldAccess(CtAnnotationFieldAccess<T>
annotationFieldAccess);
    <T> void visitCtSuperAccess(CtSuperAccess<T> f);

**Comment by GerardPaligot** [comment]
@monperrus PR ok.

File: src/test/java/spoon/reflect/visitor/processors/CheckModelProcessor.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+		if (hasScanners) {
+			checkScan(visitor, theInterface, element.getModifiers().contains(ModifierKind.ABSTRACT));
+		}
+		if (hasVisitors) {
+			checkVisit(visitor, theInterface, element.getModifiers().contains(ModifierKind.ABSTRACT));
+		}
+	}
+
+	public CheckModelProcessor hasScanners() {
```
Comment (monperrus) [inline src/test/java/spoon/reflect/visitor/processors/CheckModelProcessor.java]: method hasSomething returns a boolean by convention

File: src/test/java/spoon/reflect/visitor/processors/CheckModelProcessor.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import spoon.reflect.declaration.CtClass;
+import spoon.reflect.declaration.CtMethod;
+import spoon.reflect.declaration.CtType;
+import spoon.reflect.declaration.ModifierKind;
+import spoon.reflect.visitor.CtVisitor;
+
+import java.util.Collections;
+import java.util.List;
+
+public class CheckModelProcessor<T extends CtVisitor> extends AbstractProcessor<CtClass<?>> {
```
Comment (monperrus) [inline src/test/java/spoon/reflect/visitor/processors/CheckModelProcessor.java]: javadoc?

File: src/test/java/spoon/reflect/visitor/processors/CheckModelProcessor.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		hasScanners = true;
+		return this;
+	}
+
+	public CheckModelProcessor hasVisitors() {
+		hasVisitors = true;
+		return this;
+	}
+
+	private void checkScan(CtType<CtVisitor> scanner, CtType<Object> theInterface, boolean isAbstract) {
```
Comment (monperrus) [inline src/test/java/spoon/reflect/visitor/processors/CheckModelProcessor.java]: rename in checkPresenceofScanMethods? meaning of parameters?

File: src/test/java/spoon/reflect/visitor/CtVisitorTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import org.junit.Test;
+import spoon.Launcher;
+import spoon.reflect.visitor.processors.CheckModelProcessor;
+
+public class CtVisitorTest {
+	@Test
+	public void testMethodsInVisitor() throws Exception {
+		final Launcher launcher = new Launcher();
+		launcher.getEnvironment().setNoClasspath(true);
+		launcher.addProcessor(new CheckModelProcessor(CtVisitor.class, message -> {
```
Comment (monperrus) [inline src/test/java/spoon/reflect/visitor/CtVisitorTest.java]: what about launcher.addProcessor(new CheckThat(CtVisitor.class).hasRequiredScanners()) simply 

(the default strategy could be directly in the processor

**Comment by GerardPaligot** [comment]
@monperrus PR ok.

**Comment by GerardPaligot** [comment]
@monperrus PR ok.

**Pull request merged**